### ISSUE TITLE
2.9: Storage Pool Catch-22

### ISSUE THREAD

**Issue by neil-bowers** [lxc/lxd#2908]
Migrated from the LXC Users mailing list at the suggestion of @stgraber 

# Required information

 * Distribution: `Ubuntu Server`
 * Distribution version: `16.10`
 * The output of "lxc info" or if that fails:

```
lxc info
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFSjCCAzKgAwIBAgIRAI/DLDi2KfatUOY5stMDdCQwDQYJKoZIhvcNAQELBQAw
    NTEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEVMBMGA1UEAwwMcm9vdEBP
    bHltcHVzMB4XDTE3MDExMjAxMzIyMVoXDTI3MDExMDAxMzIyMVowNTEcMBoGA1UE
    ChMTbGludXhjb250YWluZXJzLm9yZzEVMBMGA1UEAwwMcm9vdEBPbHltcHVzMIIC
    IjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA3Zwa/ESRnqoTjLlPybEkObn+
    OBw2G3/YujTNpp6kNpX9NIr4f2KujWSCYbSL5mSX/dralQ7iKoAC2DUrIx24CKYD
    iu4ae3U338814O8BIria7JRuOO5yziw6EuKGvf+iVRYxL13NYzo0EUuW00O9Eczt
    T2F4itNned4/iQ3NTOxVaGJBDwsCKrQs0Fc4fKG6U94fIlm+r8hp3lqgojADlugC
    giGzy6fOhWvV0+ozQ3ND1qH6I0N47B7z3lBUYZv1UN52BgjoQryK4iwixQ2PDGRD
    HsjEO0KYHodxLA2r6f5Z4YQxHbxHjujUChduv69D4MTuZquoe0Qfu99/jEmOUWgG
    GW/mL+NlKEEE82Y6O97JseojF9EZUbmc0qkaPebd9uEpe44zPYZu0dQbfg9DAphB
    amviSsB1VvF5pZmlcUNnC0V/j08C4tVP0gWzwU25gtEOfOxISK+Lcw1E4Tt/9a7k
    3ZzCSb5ISf8E8IM9CgiIh+gm7/IJPwStbh7jRkJcsYj+w6pgMevgA7iwr5LBijrC
    ggjSaYRJg2jarMd/oevjVSFlpDxyFNTAV9RsKKn6LSbHgdNEbUN8Wrt8eYKkzlO+
    z2SG6FGcKsVMyIVKbZBumjiAdJIJ/ZR8DMkyra236+KWwyVrDCvbJG7uiTba9YEi
    whMwZyBdQuC3TZfJw0ECAwEAAaNVMFMwDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQM
    MAoGCCsGAQUFBwMBMAwGA1UdEwEB/wQCMAAwHgYDVR0RBBcwFYIHT2x5bXB1c4cE
    wKgBAocEwKh6ATANBgkqhkiG9w0BAQsFAAOCAgEAA+hb8AHv03bYrzQ4QzoUqLAr
    UxTxxGYK4FBYetqfiL4mJXtvp5wpf5GudlDDh7PJh217TV8b0Hhj5fZKdwn8EqGM
    /BVq7beY/5D7MEFU5k58K3teGid+CnYv0X6Ak6mIYwPW1iKiWur8jZCtSP9Mfdar
    EW6hz9TN3JwHW0c+ooBzESXkTFR8x+kffBMifub0WdY4N8dJjV2BHReTRRGOklYN
    Iv3rkYmTOKe8dH+zdcj0K7ta8RzF1C3lSZnThnxbiquBdhfEO1JLoCyC6A3Grcta
    wDLSvUV69XUgigmHNciJDxl4rl30ePDAWbkvXxIFWaSbxB6/Ct+5TnAxpMmk0bZI
    6bQc2PisFottH1zIubywAdN70w33XwpsJzKwKtEIq91HOMR7XIoB7710FIdtAZrU
    azDr0Bl+XVSyQMTh2J8txnT4QorcDs1YumAJHC/gqPMjVMhZilB5bEBKTNTdg/UT
    lY71JKABIgzGxJ+XIjLjkZui1lAQjLI+bL44kC90HZqQ+YbMKwobOyqoblE1PT27
    j+EHOw04zXUOE8MYnSrxzHQvbgIHEFKBGWGvws6F2HWxbYTuVcK00sezMPIcnFNa
    qtgCXxUqxi9gO0LemBioTEoP4URGIqZWkkdGRYf/7+d43blzHsozDIZeJ0DmhOqO
    sKiT9hVV0AHjVQeH1+g=
    -----END CERTIFICATE-----
  certificate_fingerprint: 0f059bf587989e6541aaa7750fa1cd8393ebff6b520d384a418fec12dfc1df6e
  driver: lxc
  driver_version: 2.0.7
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.8.0-37-generic
  server: lxd
  server_pid: 2926
  server_version: 2.9.1
  storage: lvm
  storage_version: 2.02.160(2) (2016-07-06) / 1.02.130 (2016-07-06) / 4.35.0
```

# Issue description

Original message to the mailing list:

Since migrating to 2.9 I've found myself in a catch-22 situation where somehow I ended up with a couple of containers that were not assigned to a storage pool upon creation.

As a side note, there should be an option to set a default storage pool value to be used with the 'lxc launch' when the -s parameter isn't specified - would've saved me a lot of time.

Consider the following:

```
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
|       NAME       |  STATE  |         IPV4         |      CREATED AT      |     LAST USED AT     | STORAGE POOL |  PROFILES   |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
| free-ray         | ERROR   |                      |                      |                      |              |             |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
| intimate-civet   | RUNNING | 192.168.1.245 (eth0) | 2017/02/19 05:24 UTC | 2017/02/19 05:24 UTC | lxc-vg       | default     |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
| minecraft-msm    | STOPPED |                      | 2017/02/19 00:33 UTC |                      | vmpool       | default-msm |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
| topical-antelope | ERROR   |                      |                      |                      |              |             |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
| warm-bluejay     | ERROR   |                      |                      |                      |              |             |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------+
```

The containers `free-ray`, `tropical-antelope` and `warm-bluejay` were created without the use of the `-s` parameter, so therefore no storage pool was assigned.

Upon trying to delete these containers:

```
lxc delete free-ray
error: No storage pool specified.
```

Not even the --force option works:

```
lxc delete free-ray --force
error: No storage pool specified.
```

After a lot of digging with the one successful container, it appears that, starting with 2.9, every container
must have a storage volume of the type 'container' with the same name as the container itself.  So I
thought, maybe I could create the storage volume and then attach it to the container, and then be able
to delete it - wrong:

```
lxc storage volume create lxc-vg container/free-ray
error: Currently not allowed to create storage volumes of type container.
```

Currently the only volume type that we're allowed to create is `custom` - so I gave that a try:

```
lxc storage volume create lxc-vg free-ray
Storage volume free-ray created
```

But then when I attempt to attach this new volume to `free-ray`, I once again encounter:

```
lxc storage volume attach lxc-vg free-ray free-ray root /
error: No storage pool specified
```
And now I'm stuck - I have three containers that can't be deleted because they're not tied to a storage pool and no way to tie them to one, and a fourth container that is tied to a now-deleted storage pool that can't be deleted either.

Short of blowing away the LXD/LXC install and starting over, is there any way to fix this?

**Comment by neil-bowers** [comment]
Response from @stgraber with my comments:

> You don't have to pass "-s" for the container to get storage associated
> with it and -s wasn't a valid option until LXD 2.9. I suspect your
> broken containers were created before the upgrade to LXD 2.9.

No, all containers were created after the upgrade to 2.9.

> The upgrade code should have updated your default profile so that this
> all works, making new container creation work and in theory should have
> also made all your existing containers keep working

This is where things may have gone wrong, as I had edited the `default` profile to include a user I wanted created for every container, along with packages, etc.  Below is the relevant section of the `default` profile:

...

devices:
  eth0:
    name: eth0
    nictype: bridged
    parent: br0
    type: nic
  root:
    path: /
    pool: ""
    type: disk
name: default
used_by:
- /1.0/containers/free-ray
- /1.0/containers/topical-antelope
- /1.0/containers/warm-bluejay
- /1.0/containers/intimate-civet
...

Note that the `pool` value is blank, and any attempt to edit it results in:

`Config parsing error: Cannot change root disk device of a profile if containers are still using it.
Press enter to open the editor again'

What is especially odd about this is that `intimate-civet` doesn't have any issues, though it was created by passing the `-s` parameter to `lxc launch`.

> Your best bet to get back to working order is:
>  - Run "lxc profile show default" and confirm that you see a root device
>    with a pool property listed there.
>  - If that's the case, then chances are your broken containers don't
>    have the default profile attached to them.
>  - Since you can't change them to fix that by either adding the default
>    profile or adding a valid root device to them, we'll need to do some DB
>    surgery:
>     - Install "sqlite3" if you don't have it already
>     - Run: sqlite3 /var/lib/lxd/lxd.db 'SELECT id FROM profiles WHERE name="default"';
>     - For each broken container, do:
>       - Run: sqlite3 /var/lib/lxd/lxd.db 'SELECT id FROM containers WHERE name="NAME"';
>       - Run: sqlite3 /var/lib/lxd/lxd.db 'INSERT INTO containers_profiles (container_id, profile_id, apply_order) VALUES (PROFILE-ID, CONTAINER-ID, 1);'

Side note: Unless there's something odd with sqlite, your insert column list and your values column list are out-of-order; when attempting this I corrected the order for my system.

Unfortunately this procedure does not work for me:

...
sqlite> select id from profiles where name = 'default';
id
39

sqlite> select id from containers where name='free-ray';
id
85

sqlite> insert into containers_profiles (container_id, profile_id, apply_order) values (85, 39, 1);
...

Then:

...
lxc info free-ray
error: No storage pool specified.
...

I believe this is because of the missing `pool` value in the default profile, which I cannot edit because there are currently containers tied to it.

**Comment by stgraber** [comment]
Yeah, the pool property being empty for that profile would totally explain this...

    sqlite3 /var/lib/lxd/lxd.db 'UPDATE profiles_devices_config SET value="default" WHERE key="pool" AND value="";'

Assuming that the database actually has the pool property set to empty, that should fix your default profile and anything that uses it.

**Comment by neil-bowers** [comment]
> Yeah, the pool property being empty for that profile would totally explain this...
> 
> `sqlite3 /var/lib/lxd/lxd.db 'UPDATE profiles_devices_config SET value="default" WHERE key="pool" AND value="";'`
>
> Assuming that the database actually has the pool property set to empty, that should fix your default profile and anything that uses it.

Ok, now we're cooking with gas:

```
└─▪ lxc-list
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
|       NAME       |  STATE  |         IPV4         |      CREATED AT      |     LAST USED AT     | STORAGE POOL |     PROFILES      |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
| free-ray         | STOPPED |                      | 2017/02/19 04:26 UTC |                      | lxc-vg       | default           |
|                  |         |                      |                      |                      |              | default-29        |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
| intimate-civet   | RUNNING | 192.168.1.245 (eth0) | 2017/02/19 05:24 UTC | 2017/02/19 05:24 UTC | lxc-vg       | default           |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
| minecraft-msm    | STOPPED |                      | 2017/02/19 00:33 UTC |                      | vmpool       | default-msm       |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
| topical-antelope | STOPPED |                      | 2017/02/19 05:18 UTC |                      | lxc-vg       | default           |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
| warm-bluejay     | STOPPED |                      | 2017/02/19 05:23 UTC |                      | lxc-vg       | default           |
+------------------+---------+----------------------+----------------------+----------------------+--------------+-------------------+
```

And `lxc info` doesn't fail now:

```
lxc info free-ray
Name: free-ray
Remote: unix:/var/lib/lxd/unix.socket
Architecture: x86_64
Created: 2017/02/19 04:26 UTC
Status: Stopped
Type: persistent
Profiles: default, default-29
```

However, I'm still not able to delete the defective containers:

```
lxc delete free-ray
error: not found
```

My guess is the `error: not found` is referring to a lack of a container volume assigned to the container:

```
lxc storage volume list lxc-vg
+-----------+------------------------------------------------------------------+---------+
|   TYPE    |                               NAME                               | USED BY |
+-----------+------------------------------------------------------------------+---------+
| container | intimate-civet                                                   | 0       |
+-----------+------------------------------------------------------------------+---------+
| custom    | free-ray                                                         | 0       |
+-----------+------------------------------------------------------------------+---------+
| image     | 87a1f1c305615024124b238c0fcbed99c11193cab4bbc5c340ca2500c99b1b1f | 0       |
+-----------+------------------------------------------------------------------+---------+
```

Attempting to attach the previously created `custom` volume as the root disk results in:

```
lxc storage volume attach lxc-vg free-ray free-ray root /
error: Root disk entry may not have a "source" property set.
```

My day job as a DBA suggests that editing the database and removing all references to the broken containers might fix the issue, but I don't know enough about the internal code to risk it at the moment.

**Comment by stgraber** [comment]
Hmm, well, first lets see what actually exists on disk, then we can make the database act like it :)

What does "lvs" show?

**Comment by brauner** [comment]
What @stgraber said. In case the container actually exists on disk the way to fix the db would be:

- First check if there is a storage volume for that container:
```
sqlite3 /var/lib/lxd/lxd.db 'SELECT name FROM storage_volumes'
```

If the container you're trying to delete does not show up but actually exists, you can create a storage volume for the container in the database via:

First get the ID of the storage pool via:
```
sqlite3 'SELECT id FROM storage_pools WHERE name=<pool-name>'
```
Then create storage volume for the container in the db using the id for the pool you got before:
```
sqlite3 'INSERT INTO storage_volumes (storage_pool_name, type, name) VALUES (<pool-id>, 0, <container-name>)'
```

**Comment by stgraber** [comment]
If you really don't care about any of those containers you should:
 - Remove any LV that's listed for them in "lvs"
 - 'DELETE FROM containers WHERE name IN ("free-ray", "minecraft-msm", "topical-antelope", "warm-blueyay");'
 - Confirm that there's nothing for those containers in: 'SELECT * FROM storage_volumes;' if there is, you'll want to DELETE those rows
 - Do a "find /var/lib/lxd | grep NAME" for each of them and remove anything that it finds

That should take care of completely wiping those containers and let you reuse their name in the future without getting weird collisions.

**Comment by neil-bowers** [comment]
> Hmm, well, first lets see what actually exists on disk, then we can make the database act like it :)
> 
> What does "lvs" show?

```
  LV                                                                      VG         Attr       LSize   Pool        Origin                                                $
  root                                                                    Olympus-vg -wi-ao---- 106.75g                                                                   $
  swap_1                                                                  Olympus-vg -wi-ao----  12.00g                                                                   $
  LXDThinpool                                                             lxc-vg     twi-aotz-- 277.39g                                                                   $
  containers_accurate--turtle                                             lxc-vg     Vwi-aotz--  10.00g LXDThinpool images_87a1f1c305615024124b238c0fcbed99c11193cab4bbc5c$
  containers_intimate--civet                                              lxc-vg     Vwi-aotz--  10.00g LXDThinpool images_87a1f1c305615024124b238c0fcbed99c11193cab4bbc5c$
  custom_free-ray                                                         lxc-vg     Vwi-aotz--  10.00g LXDThinpool                                                       $
  images_87a1f1c305615024124b238c0fcbed99c11193cab4bbc5c340ca2500c99b1b1f lxc-vg     Vwi-a-tz--  10.00g LXDThinpool 
```

So the container for `free-ray` doesn't actually exist on disk, just the custom one I created when trying to work around the issue.

>First check if there is a storage volume for that container:
>`sqlite3 /var/lib/lxd/lxd.db 'SELECT name FROM storage_volumes'`

<table border="1">
<tr><th>id</th><th>name</th><th>storage_pool_id</th><th>type</th></tr>
<tr><td>11</td><td>intimate-civet</td><td>6</td><td>0</td></tr>
<tr><td>12</td><td>87a1f1c305615024124b238c0fcbed99c11193cab4bbc5c340ca2500c99b1b1f</td><td>6</td><td>1</td></tr>
<tr><td>13</td><td>free-ray</td><td>6</td><td>2</td></tr>
<tr><td>14</td><td>accurate-turtle</td><td>6</td><td>0</td></tr>
</table>

So there is a storage volume for `free-ray`, but the type is incorrect assuming that a 0 represents a `container` type and a 2 represents a `custom` type - would updating this row in the database work?

**Comment by stgraber** [comment]
Not really. You'd also need to rename the LV, create the paths in /var/lib/lxd/storage-pools and you'd end up with a container that looks valid but has nothing in its root filesystem so it wouldn't boot.

It's best to just remove that storage volume with "lxc storage volume delete default free-ray", get rid of the container entries and then create it again once things are clean, using your fixed default profile.

**Comment by neil-bowers** [comment]
>  get rid of the container entries and then create it again once things are clean, using your fixed default profile.

Just so I'm clear - identify the container ID, and then go through each of the tables and delete all rows associated with that container ID?

**Comment by stgraber** [comment]
DELETE can cascade to do that for you, but when using the sqlite3 shell, you'll need to first run:

    PRAGMA foreign_keys=ON;

And then any DELETE you do in the containers table will cascade and delete any related row in the other containers_* tables.

**Comment by neil-bowers** [comment]
>And then any DELETE you do in the containers table will cascade and delete any related row in the other containers_* tables.

Fantastic - appears to have worked perfectly!

One last question - is there a way to regenerate the default profile from the base install?

**Comment by stgraber** [comment]
The default profile, post "lxd init" on LXD 2.9, should look like:

```
config: {}
description: Default LXD profile
devices:
  eth0:
    nictype: bridged
    parent: lxdbr0
    type: nic
  root:
    path: /
    pool: default
    type: disk
name: default
```

**Comment by stgraber** [comment]
This issue should now have been resolved in git and in LXD 2.9.2, closing.

If anyone who has installed LXD 2.9.2 runs into something like this, please file another issue.
